common/middleware/src/line_rules.rs

```diff
@@ -0,0 +1,260 @@
+use crate::{Middleware, Status};
+use http::types::body::LineMetaMut;
+use regex::bytes::Regex as BytesRegex;
+use regex::Regex as TextRegex;
+use thiserror::Error;
+
+static REDACTED_TEXT: &str = "[REDACTED]";
+static REDACT_BYTES: &[u8] = &REDACTED_TEXT.as_bytes();
+
+pub struct LineRules {
+    exclusion_text: Vec<TextRegex>,
+    exclusion_bytes: Vec<BytesRegex>,
+    inclusion_text: Vec<TextRegex>,
+    inclusion_bytes: Vec<BytesRegex>,
+    redact_text: Vec<TextRegex>,
+    redact_bytes: Vec<BytesRegex>,
+}
+
+#[derive(Clone, Debug, Error)]
+pub enum LineRulesError {
+    #[error(transparent)]
+    RegexError(regex::Error),
+}
+
+impl LineRules {
+    pub fn new(
+        exclusion: &[String],
+        inclusion: &[String],
+        redact: &[String],
+    ) -> Result<LineRules, LineRulesError> {
+        Ok(LineRules {
+            exclusion_text: map_to_regex(exclusion)?,
+            inclusion_text: map_to_regex(inclusion)?,
+            redact_text: map_to_regex(redact)?,
+            // At this time, regex values has been validated, it's safe to unwrap()
+            exclusion_bytes: exclusion
+                .iter()
+                .map(|s| BytesRegex::new(s).unwrap())
+                .collect(),
+            inclusion_bytes: inclusion
+                .iter()
+                .map(|s| BytesRegex::new(s).unwrap())
+                .collect(),
+            redact_bytes: redact.iter().map(|s| BytesRegex::new(s).unwrap()).collect(),
+        })
+    }
+
+    /// Applies inclusion and exclusion rules and replaces the redacted values
+    fn process_text<'a>(&self, line: &'a mut dyn LineMetaMut) -> Status<&'a mut dyn LineMetaMut> {
+        let value = line.get_line().0.unwrap();
+
+        // If it doesn't match any inclusion rule -> skip
+        if !self.inclusion_text.is_empty() && !self.inclusion_text.iter().any(|r| r.is_match(value))
+        {
+            return Status::Skip;
+        }
+
+        // If any exclusion rule matches -> skip
+        if self.exclusion_text.iter().any(|r| r.is_match(value)) {
+            return Status::Skip;
+        }
+
+        if !self.redact_text.is_empty() {
+            let mut v = value.to_string();
+            for r in self.redact_text.iter() {
```

Comment:
If we went with that method we could replace the Vec<Regex>'s with the regex crate's RegexSet https://docs.rs/regex/1.4.5/regex/struct.RegexSet.html which does a single pass match on all of the patterns contained in it, rather than having to match against every Regex separately

Reply:
I think the logic would be more complex than that, like:

- Check each match of each regex
- Compare boundaries of each match, expanding when necessary
- Rebuild a new line with substrings of the original line and the redacted portions.

That compared to a loop of regex replace calls.

Looking at the docs for `RegexSet`, I think it only yields which regex matched (not the start and end indexes of the match). 

From the user pov, we can document the limitations of applying the redaction in the defined order and recommend the user to define the patterns in order, from more to less specific, wdyt?

Reply:
> Looking at the docs for RegexSet, I think it only yields which regex matched (not the start and end indexes of the match).

Ahh, darn, I've only ever used it to test for membership, pity it can't return locations.

> From the user pov, we can document the limitations of applying the redaction in the defined order and recommend the user to define the patterns in order, from more to less specific, wdyt?

I think we should do our best to handle the overlapping cases and try to reduce the chances of it surprising anyone as much as possible. There are some cases where it would be impossible to order the regexes so that bot applied correctly, eg in the case where the suffix of one pattern was the same as the prefix of another only one could possibly be applied correctly.

The algorithm for combining overlapping matches shouldn't be too bad, something like the following pseudocode:

```
let matches: Vec<(usize, usize)> = vec![];

for pattern in patterns.iter() {
    matches.extend(pattern.match(input);
}

let matches  = matches.sort_unstable();
let new_matches = vec![];

let mut current: Option<(usize, usize)> = None;
for i in (0..matches.len()) {
    let (start, mut end) = if let Some(start, end) = current {
        (start, end)
    } else {
        (matches[i].0, matches[i].1) 
    }

    if matches[i + 1].0 < end {
        // next match overlaps this one, extend the end of current
        end = matches[i + 1].1
    } else {
        // next match doesn't overlap, current match is complete, push it
        new_matches.push(current.take().unwrap())
    }
    
}

return new_matches
```

That's some pretty rought pseudocode and I'm pretty sure it could be condensed a bit, but I think it's the right algorithm

Reply:
I'll update the logic to handle overlapping matches gracefully 👍